### PR TITLE
Qdb improvements

### PR DESCRIPTION
**Pull request by ucgJhe**
1. minor bugs fixed
2. minor refactored
3. add SnapshotManager for both commands step-in and step-over

TODO:
1. add test for all commands of qdb
2. improve memory access functionality qdb

### CODE REVIEW

File: qiling/debugger/qdb/qdb.py [hunk 1]
```diff
@@ -178,92 +204,74 @@ def do_run(self: QlQdb, *args) -> None:
...
-            else:
-                self._run(count=1)
+        if address is None:
+            address = self.cur_addr
 
-                if self.rr:
-                    # context save after execution, so we could do diff
-                    # on two states before and after.
-                    self.rr.save()
+        print(f"{color.CYAN}continued from 0x{address:08x}{color.END}")
```
Comment (elicn) [inline qiling/debugger/qdb/qdb.py:256]: There are a lot of color printing here; consider creating a method to print out colorful messages by specifying the message type (e.g. `qdb_print(msgtype: QDB_MSG, msg: str)` , which may be called: `qdb_print(QDB_MSG.WARNING, "this is a warning!")` )

File: qiling/debugger/qdb/qdb.py [hunk 2]
```diff
@@ -305,17 +303,15 @@ def do_breakpoint(self: QlQdb, address: Optional[int] = 0) -> None:
         print(f"{color.CYAN}[+] Breakpoint at 0x{address:08x}{color.END}")
 
     @parse_int
-    def do_continue(self: QlQdb, address: Optional[int] = 0) -> None:
+    def do_disassemble(self: QlQdb, address: Optional[int] = 0, *args) -> None:
```
Comment (elicn) [inline qiling/debugger/qdb/qdb.py:306]: What is the `*args` for here (and in the following methods)?
A well-defined API usually doesn't need that.

File: qiling/debugger/qdb/render/render.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+from ..context import Context
+from ..const import color
+
+
+def get_terminal_size() -> Iterable:
+    """
+    get terminal window height and width
+    """
+
+    return map(int, os.popen('stty size', 'r').read().split())
```
Comment (elicn) [inline qiling/debugger/qdb/render/render.py:19]: Use `os.get_terminal_size` instead.
Much safer and more reliable.

File: qiling/debugger/qdb/render/render.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+                line = f"{color.UNDERLINE}{color.BOLD}{line}"
+
+            if idx % self.regs_a_row == 0 and idx != 32:
+                line += "\n"
+
+            lines += line
+
+        print(lines.format(*regs.values()))
+
+    def render_stack_dump(self, arch_sp: int, pointer_size: int = 4) -> None:
```
Comment (elicn) [inline qiling/debugger/qdb/render/render.py:86]: Do not use defaults that fits only some of the class derivatives (i.e. `pointer_size = 4`)

File: qiling/debugger/qdb/render/render.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        """
+        helper function for redering stack dump
+        """
+
+        for idx in range(self.stack_num):
+            addr = arch_sp + idx * pointer_size
+            if (val := self.try_read_pointer(addr)[0]):
```
Comment (elicn) [inline qiling/debugger/qdb/render/render.py:93]: This is C code written in Python syntax..
Avoid assigning variables with values and checking them in a single statement.

File: qiling/debugger/qdb/render/render.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        if (cur_insn := lines.get("current", None)):
+            prophecy = self.predictor.predict()
+            self.print_asm(cur_insn, to_jump=prophecy.going)
+
+        # assembly after current location
+        if (forward := lines.get("forward", None)):
+            for line in forward:
+                self.print_asm(line)
+
+    def swap_reg_name(self, cur_regs: Mapping["str", int], extra_dict=None):
```
Comment (elicn) [inline qiling/debugger/qdb/render/render.py:131]: Why the `str` type is surrounded with qoutes..?

File: qiling/debugger/qdb/render/render.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        return cur_regs 
+
+    def print_asm(self, insn: CsInsn, to_jump: bool = False) -> None:
+        """
+        helper function for printing assembly instructions, indicates where we are and the branch prediction
+        provided by BranchPredictor
+        """
+
+        opcode = "".join(f"{b:02x}" for b in insn.bytes)
+        trace_line = f"0x{insn.address:08x} │ {opcode:15s} {insn.mnemonic:10} {insn.op_str:35s}"
```
Comment (elicn) [inline qiling/debugger/qdb/render/render.py:150]: Assembly line format could differ quite a lot between architectures.
Consider having these values as inherited class members that change from arch to arch.

**Review by elicn — APPROVED** [review]

File: qiling/debugger/qdb/context.py [hunk 8]
```diff
@@ -68,7 +73,7 @@ def try_read_pointer(self, address: int) -> Optional[bytes]:
         try to read pointer size of data from ql.mem
         """
 
-        return self.try_read(address, self.ql.pointersize)
+        return self.try_read(address, self.pointersize)
```
Comment (elicn) [inline qiling/debugger/qdb/context.py:76]: I think that the size of the read should be the longest possible instruction for that specific arch (i.e. 4 for ARM, 15 for Intel, etc.)

**Review by xwings — APPROVED** [review]

**Pull request merged**